implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+        SecretAttributes::Buffer(SHA256_SIZE_U32)
+    }
+
+    /// Secret attributes for the k key
+    fn k_attributes() -> SecretAttributes {
+        SecretAttributes::Aes256
+    }
+
+    /// Read the message 1 payload which is present after the public key
+    fn read_message1_payload(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_end(message, Self::key_size())
+    }
+
+    /// Read the message 2 encrypted key, which is present after the public key
+    fn read_message2_encrypted_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_middle(message, Self::key_size(), Self::encrypted_key_size())
+    }
+
+    /// Read the message 2 encrypted payload, which is present after the encrypted key
+    fn read_message2_payload(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_end(message, Self::key_size() + Self::encrypted_key_size())
+    }
+
+    /// Read the message 3 encrypted key at the beginning of the message
+    fn read_message3_encrypted_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_start(message, Self::encrypted_key_size())
+    }
+
+    /// Read the message 3 payload which is present after the encrypted key
+    fn read_message3_payload(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_end(message, Self::encrypted_key_size())
+    }
+
+    /// Read the first 'length' bytes of the message
+    fn read_start(message: &Vec<u8>, length: usize) -> Result<&[u8]> {
+        if message.len() < length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[0..length])
+    }
+
+    /// Read the bytes of the message after the first 'drop_length' bytes
+    fn read_end(message: &Vec<u8>, drop_length: usize) -> Result<&[u8]> {
+        if message.len() < drop_length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[drop_length..])
+    }
+
+    /// Read 'length' bytes of the message after the first 'drop_length' bytes
+    fn read_middle(message: &Vec<u8>, drop_length: usize, length: usize) -> Result<&[u8]> {
+        if message.len() < drop_length + length {
+            return Err(XXError::MessageLenMismatch.into());
+        }
+        Ok(&message[drop_length..(drop_length + length)])
+    }
+
+    /// Read the bytes of a key at the beginning of a message
+    fn read_key(message: &Vec<u8>) -> Result<&[u8]> {
+        Self::read_start(message, Self::key_size())
+    }
+
+    /// Size of a public key
+    fn key_size() -> usize {
+        CURVE25519_PUBLIC_LENGTH_USIZE
+    }
+
+    /// Size of an encrypted key
+    fn encrypted_key_size() -> usize {
+        Self::key_size() + AES_GCM_TAGSIZE_USIZE
+    }
+}
+
+/// The `HandshakeState` contains all the variables necessary to follow the Noise protocol
+#[derive(Debug, Clone)]
+pub(super) struct HandshakeState {
+    pub(super) s: KeyId,
+    e: KeyId,
+    k: KeyId,
+    re: PublicKey,
+    pub(super) rs: PublicKey,
+    n: u16,
```

Comment:
According to noise spec `n` should be u64. While during the handshake `n` value could be easily fitted into `u8`. Where `u16` comes from?